### ISSUE TITLE
wc_get_products() flexibility

### ISSUE THREAD

**Issue by leewillis77** [woocommerce/woocommerce#12961]
## EXPLANATION OF THE ISSUE

Under WC 2.6, you can run a query like this:

```php
$args = array(
	'post_type'   => 'product',
	'meta_key'    => 'total_sales',
	'orderby'     => 'meta_value_num',
	'order'       => 'DESC',
	'numberposts' => 2,
	'meta_query' => array(
		array(
			'key' => '_thumbnail_id',
			'compare' => 'EXISTS'
		),
	)
);
$products = get_posts( $args );
```

This will fetch the 2 highest selling products that have featured images. This is made possible by:

a) the ability to construct a meta_query (`'meta_query' => array(...)`)
b) the ability to order results by the value of a meta field (`'meta_key'    => 'total_sales', 'orderby'     => 'meta_value_num'`)

As far as I can tell, neither of these are supported by wc_get_products():

a) meta_query is always set to an empty array (https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-product-data-store-cpt.php#L1031)
b) There's no way of passing in meta_key to the query.

I haven't followed all the way through to check if I can hijack with a filter to add these in later, and I understand that supporting the current facilities as-is probably isn't a goal of the current abstraction work.

However these seem like reasonable examples of what might be required in terms of querying products (ordering by arbitrary product properties, and requesting only products that have a specific property set) - while the example I've given above may seem a bit odd, it's a real-world piece of code.

Opening this to see if this sort of query will be supported in 2.7 or beyond.

**Comment by justinshreve** [comment]
We have plans to introduce a more robust way to query data without using `get_posts` in 2.8: https://github.com/woocommerce/woocommerce/issues/12677.

We'll have something like `WC_Products_Query` and `WC_Orders_Query` (think WP_Query or WP_User_Query) that will support things like custom meta, the other props, ordering, etc.

In the meantime, `wp_get_products` can be used for simpler queries, and `get_posts`/`WP_Query` can be used for more complex queries. We'll plan on having a guide to move to newer code like we are doing for the 2.7 compatibility updates.

I'm going to close this issue since we have the other issue open -- but please share any other notes/requests/or real world examples that we can use to design the new queries like you have above!

**Comment by leewillis77** [comment]
Awesome response - thanks Justin!

**Comment by webdados** [comment]
@leewillis77 It's a good thing they're planning on do this the right way. I just think there should be efforts to ship this with 2.7, when the data abstraction is happening.

Right now, WooCommerce itself will not (yet) move the orders away from post and post_meta, so we can still use WordPress own functions to get orders / products based on meta, but the data abstraction layer that CRUD implements will allow developers to actually move the orders / products to other tables or even external systems, and losing the ability to search by meta seems like a big deal to me.

Anyway, I just hope it gets done ASAP :-) I've got at least two plugins on the repository, used by thousands, that need to get orders based on meta, and I'm pretty sure I'm not alone.

**Comment by leewillis77** [comment]
> the data abstraction layer that CRUD implements will allow developers to actually move the orders / products to other tables or even external systems,

I can't see anyone *actually* doing that though. As far as I'm concerned it seems clear that WP_Query, and post_meta are still supported through the 2.7 cycle. I'd agree that stuff shouldn't be moved out of posts / postmeta without this being in place.

**Comment by webdados** [comment]
> I'd agree that stuff shouldn't be moved out of posts / postmeta without this being in place.

I'd go further and say that this must be put in place way before anything moves out of posts and post_meta. I'm afraid that because this is set to be implemented only in 2.8, and not in 2.7, there's not enough time for developers to, first understand they need to update code and, second actually implement the changes.

**Comment by leewillis77** [comment]
This was closed in favour of #12677, and then split out into #13645 which was then closed by the PR In #16291. 

That provides a link to the wiki at https://github.com/woocommerce/woocommerce/wiki/wc_get_products-and-WC_Product_Query

However, having read that, and looking at the current proposal to move to the custom table implementation, I can't see that the original requirements raised in this bug report have actually been met. 

Namely, I don't believe it's currently possible to write a call to wc_get_products() that will do the following consistently across CPT-storage and custom table-storage:
- return results sorted by a specified meta key
- filter the results by the existence of a specified meta key

I'm not even sure that can be achieved with storage-specific code without resorting to intercepting and manipulating the SQL itself?
